From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ingilying <ingilying@outlook.com>
Date: Thu, 30 Jul 2026 23:06:54 +0800
Subject: [PATCH] Tumble: Add server brand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index d9127bc6c5801521a5dd1171e4e215f241934034..f59c88c9799bd7c61548c706642c8d867afb78dc 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -28,6 +28,13 @@ public interface ServerBuildInfo {
     Key BRAND_LOPHINE_ID = Key.key("lophinelabs", "lophine");
     // Lophine end
 
+    // Tumble start
+    /**
+     * The brand id for Tumble.
+     */
+    Key BRAND_TUMBLE_ID = Key.key("org.tumblemc", "tumble");
+    // Tumble end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
