--- a/lophine-server/build.gradle.kts
+++ b/lophine-server/build.gradle.kts
@@ -50,7 +_,23 @@
         }
     }
 
-    activeFork = lophine
+    val tumble = forks.register("tumble") {
+        forks = lophine
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = lophine.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("tumble-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+
+        upstream.patchDir("lophineServer") {
+            upstreamPath = "lophine-server"
+            excludes = setOf("src/minecraft", "paper-patches", "folia-patches", "minecraft-patches", "build.gradle.kts", "build.gradle.kts.patch")
+            patchesDir = rootDirectory.dir("tumble-server/lophine-patches")
+            outputDir = rootDirectory.dir("lophine-server")
+        }
+    }
+
+    activeFork = tumble
 
 
     updatingMinecraft {
@@ -124,12 +_,16 @@
         resources { srcDir("../paper-server/src/main/resources") }
         java { srcDir("../folia-server/src/main/java") }
         resources { srcDir("../folia-server/src/main/resources") }
+        java { srcDir("../lophine-server/src/main/java") }
+        resources { srcDir("../lophine-server/src/main/resources") }
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
         java { srcDir("../folia-server/src/test/java") }
         resources { srcDir("../folia-server/src/test/resources") }
+        java { srcDir("../lophine-server/src/test/java") }
+        resources { srcDir("../lophine-server/src/test/resources") }
     }
 }
 val log4jPlugins = sourceSets.create("log4jPlugins") {
@@ -157,7 +_,7 @@
 }
 
 dependencies {
-    implementation(project(":lophine-api")) // Lophine - api update
+    implementation(project(":tumble-api")) // Tumble - API
     // Lophine start - Dependenices insert
     implementation("net.objecthunter:exp4j:0.4.8")
     implementation("com.electronwill.night-config:toml:3.8.4") // Night config
@@ -233,14 +_,14 @@
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Lophine", // Lophine - Rebrand
+            "Implementation-Title" to "Tumble",
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "Lophine", // Lophine - Rebrand
+            "Specification-Title" to "Tumble",
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "LophineLabs", // Lophine - Rebrand
-            "Brand-Id" to "lophinelabs:lophine", // Lophine - Rebrand
-            "Brand-Name" to "Lophine", // Lophine - Rebrand
+            "Specification-Vendor" to "TumbleMC",
+            "Brand-Id" to "org.tumblemc:tumble",
+            "Brand-Name" to "Tumble",
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
