From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NatJerry <xiaomajunqwq123@outlook.com>
Date: Sat, 1 Aug 2026 11:43:57 +0800
Subject: [PATCH] Add Purpur Anvil Modify


diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index fdd03c9d73a8a0527d0ead3fbbba20083c617e2c..6b3bf9a7e9250e69fcac55940b55e22533d59781 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -64,6 +64,9 @@ public abstract class AbstractContainerMenu {
     private final List<ContainerListener> containerListeners = Lists.newArrayList();
     private @Nullable ContainerSynchronizer synchronizer;
     private boolean suppressRemoteUpdates;
+    // Tumble start // Purpur - Anvil API
+    @Nullable protected ItemStack activeQuickItem = null;
+    // Tumble end
     // CraftBukkit start
     public boolean checkReachable = true;
     public abstract org.bukkit.inventory.InventoryView getBukkitView();
diff --git a/net/minecraft/world/inventory/AnvilMenu.java b/net/minecraft/world/inventory/AnvilMenu.java
index a577225f59bb562b860cb90db72b4d658a69956f..c01e92b8831ab40255dfbcf9ac68b2771d3ec698 100644
--- a/net/minecraft/world/inventory/AnvilMenu.java
+++ b/net/minecraft/world/inventory/AnvilMenu.java
@@ -5,6 +5,8 @@ import it.unimi.dsi.fastutil.objects.Object2IntMap.Entry;
 import net.minecraft.core.Holder;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.network.chat.Component;
+import net.minecraft.network.protocol.game.ClientboundContainerSetDataPacket;
+import net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.util.Mth;
@@ -50,6 +52,12 @@ public class AnvilMenu extends ItemCombinerMenu {
     public int maximumRepairCost = 40;
     private org.bukkit.craftbukkit.inventory.view.CraftAnvilView bukkitEntity;
     // CraftBukkit end
+
+    // Tumble start - Purpur Anvil API
+    public boolean bypassCost = false;
+    public boolean canDoUnsafeEnchants = false;
+    // Tumble end - Purpur Anvil API
+
     public boolean bypassEnchantmentLevelRestriction = false; // Paper - bypass anvil level restrictions
 
     public AnvilMenu(final int containerId, final Inventory inventory) {
@@ -81,7 +89,12 @@ public class AnvilMenu extends ItemCombinerMenu {
 
     @Override
     protected void onTake(final Player player, final ItemStack carried) {
+        // Tumble start - Purpur Anvil API
+        ItemStack itemstack = this.activeQuickItem != null ? this.activeQuickItem : carried;
+        if (org.purpurmc.purpur.event.inventory.AnvilTakeResultEvent.getHandlerList().getRegisteredListeners().length > 0) new org.purpurmc.purpur.event.inventory.AnvilTakeResultEvent(player.getBukkitEntity(), getBukkitView(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack)).callEvent();
+        // Tumble end - Purpur Anvil API
         if (!player.hasInfiniteMaterials()) {
+            if (this.bypassCost) ((ServerPlayer) player).lastSentExp = -1; else // Tumble - Purpur Anvil API
             player.giveExperienceLevels(-this.cost.get());
         }
 
@@ -134,13 +147,19 @@ public class AnvilMenu extends ItemCombinerMenu {
 
     @Override
     public void createResult() {
+        // Tumble start - Purpur Anvil API
+        this.bypassCost = false;
+        this.canDoUnsafeEnchants = false;
+        if (org.purpurmc.purpur.event.inventory.AnvilUpdateResultEvent.getHandlerList().getRegisteredListeners().length > 0) new org.purpurmc.purpur.event.inventory.AnvilUpdateResultEvent(getBukkitView()).callEvent();
+        // Tumble end - Purpur Anvil API
+
         ItemStack input = this.inputSlots.getItem(0);
         this.onlyRenaming = false;
         this.cost.set(1);
         int price = 0;
         long tax = 0L;
         int namingCost = 0;
-        if (!input.isEmpty() && EnchantmentHelper.canStoreEnchantments(input)) {
+        if (!input.isEmpty() && this.canDoUnsafeEnchants || EnchantmentHelper.canStoreEnchantments(input)) { // Tumble - Purpur Anvil API
             ItemStack result = input.copy();
             ItemStack addition = this.inputSlots.getItem(1);
             ItemEnchantments.Mutable enchantments = new ItemEnchantments.Mutable(EnchantmentHelper.getEnchantmentsForCrafting(result));
@@ -198,23 +217,38 @@ public class AnvilMenu extends ItemCombinerMenu {
                         int level = entry.getIntValue();
                         level = current == level ? level + 1 : Math.max(level, current);
                         Enchantment enchantment = enchantmentHolder.value();
-                        boolean compatible = enchantment.canEnchant(input);
+                        // Tumble start - Config to allow unsafe enchants
+                        boolean compatible = this.canDoUnsafeEnchants ||
+                                org.tumblemc.tumble.config.modules.function.AnvilConfig.allow_inapplicable_enchants ||
+                                enchantment.canEnchant(input); // whether the enchantment can be applied on specific item type
+                        boolean canEnchant1 = true; // whether two incompatible enchantments can be applied on a single item
+                        // Tumble end - Config to allow unsafe enchants
                         if (this.player.hasInfiniteMaterials() || input.is(Items.ENCHANTED_BOOK)) {
                             compatible = true;
                         }
 
+                        java.util.Set<Holder<Enchantment>> removedEnchantments = new java.util.HashSet<>(); // Tumble - Config to allow unsafe enchants
                         for (Holder<Enchantment> other : enchantments.keySet()) {
                             if (!other.equals(enchantmentHolder) && !Enchantment.areCompatible(enchantmentHolder, other)) {
-                                compatible = false;
+                                canEnchant1 = this.canDoUnsafeEnchants || org.tumblemc.tumble.config.modules.function.AnvilConfig.allow_incompatible_enchants
+//                                        || org.purpurmc.purpur.PurpurConfig.allowIncompatibleEnchants
+                                ; // Tumble - Purpur Anvil API // Tumble - canEnchant -> canEnchant1 - Config to allow unsafe enchants
+                                // Tumble start - Config to allow unsafe enchants
+                                if (!canEnchant1 && org.tumblemc.tumble.config.modules.function.AnvilConfig.replace_incompatible_enchants) {
+                                    removedEnchantments.add(other);
+                                    canEnchant1 = true;
+                                }
+                                // Tumble end - Config to allow unsafe enchants
                                 price++;
                             }
                         }
+                        enchantments.removeIf(removedEnchantments::contains); // Tumble - Config to allow unsafe enchants
 
-                        if (!compatible) {
+                        if (!compatible || !canEnchant1) { // Tumble - Config to allow unsafe enchants
                             isAnyEnchantmentNotCompatible = true;
                         } else {
                             isAnyEnchantmentCompatible = true;
-                            if (level > enchantment.getMaxLevel() && !this.bypassEnchantmentLevelRestriction) { // Paper - bypass anvil level restrictions
+                            if (!org.tumblemc.tumble.config.modules.function.AnvilConfig.allow_higher_enchants_levels && level > enchantment.getMaxLevel() && !this.bypassEnchantmentLevelRestriction) { // Paper - bypass anvil level restrictions // Tumble - Config to allow unsafe enchants
                                 level = enchantment.getMaxLevel();
                             }
 
@@ -267,6 +301,12 @@ public class AnvilMenu extends ItemCombinerMenu {
                 this.onlyRenaming = true;
             }
 
+            // Tumble start - Purpur Anvil API
+            if (this.bypassCost && this.cost.get() >= this.maximumRepairCost) {
+                this.cost.set(this.maximumRepairCost - 1);
+            }
+            // Tumble end - Purpur Anvil API
+
             if (this.cost.get() >= this.maximumRepairCost && !this.player.hasInfiniteMaterials()) { // CraftBukkit
                 result = ItemStack.EMPTY;
             }
@@ -287,6 +327,14 @@ public class AnvilMenu extends ItemCombinerMenu {
 
             org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareAnvilEvent(this.getBukkitView(), result); // CraftBukkit
             this.broadcastChanges();
+
+            // Tumble start - Purpur Anvil API
+            if ((this.canDoUnsafeEnchants || org.tumblemc.tumble.config.modules.function.AnvilConfig.allow_inapplicable_enchants || org.tumblemc.tumble.config.modules.function.AnvilConfig.allow_incompatible_enchants
+            ) && result != ItemStack.EMPTY) { // Tumble - Config to allow unsafe enchants
+                ((ServerPlayer) this.player).connection.send(new ClientboundContainerSetSlotPacket(this.containerId, this.incrementStateId(), 2, result));
+                ((ServerPlayer) this.player).connection.send(new ClientboundContainerSetDataPacket(this.containerId, 0, this.cost.get()));
+            }
+            // Tumble end - Purpur Anvil API
         } else {
             org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareAnvilEvent(this.getBukkitView(), ItemStack.EMPTY); // CraftBukkit
             this.cost.set(AnvilMenu.DEFAULT_DENIED_COST); // CraftBukkit - use a variable for set a cost for denied item
