From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NatJerry <xiaomajunqwq123@outlook.com>
Date: Sat, 1 Aug 2026 13:29:56 +0800
Subject: [PATCH] Tumble: Rebrand server


diff --git a/src/main/java/me/earthme/luminol/config/modules/misc/ServerModNameConfig.java b/src/main/java/me/earthme/luminol/config/modules/misc/ServerModNameConfig.java
index 347f9462a97d2b150730c81d4fbded57291802b9..7ed5c081433b0193ecaee20ad3cc7277e0e2c37d 100644
--- a/src/main/java/me/earthme/luminol/config/modules/misc/ServerModNameConfig.java
+++ b/src/main/java/me/earthme/luminol/config/modules/misc/ServerModNameConfig.java
@@ -7,7 +7,7 @@ import me.earthme.luminol.enums.EnumConfigCategory;
 @ConfigClassInfo(category = EnumConfigCategory.MISC, name = "server_mod_name")
 public class ServerModNameConfig {
     @ConfigInfo(name = "name")
-    public static String serverModName = "Lophine";
+    public static String serverModName = "Tumble"; //Tumble - Rebrand to Tumble
 
     @ConfigInfo(name = "vanilla_spoof")
     public static boolean fakeVanilla = false;
