From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NatJerry <xiaomajunqwq123@outlook.com>
Date: Sat, 1 Aug 2026 13:30:07 +0800
Subject: [PATCH] Add Tumble Config


diff --git a/src/main/java/me/earthme/luminol/config/ConfigManager.java b/src/main/java/me/earthme/luminol/config/ConfigManager.java
index 678380847d05f2c784eb262faea75470fde4b45a..2aa665a7cc8780d02e6b415b6f544c66ec7e1fdb 100644
--- a/src/main/java/me/earthme/luminol/config/ConfigManager.java
+++ b/src/main/java/me/earthme/luminol/config/ConfigManager.java
@@ -35,6 +35,7 @@ public class ConfigManager {
                 "fun.bm.lophine.carpet.config.modules"
         ));
         // Lophine end - add carpet config
+        registerConfig("tumble",builder.of("tumble", "org.tumblemc.tumble.config.modules")); //Tumble - add tumble config
         preLoad();
     }
 
