From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NatJerry <xiaomajunqwq123@outlook.com>
Date: Sat, 1 Aug 2026 13:30:07 +0800
Subject: [PATCH] Add Tumble Config


diff --git a/src/main/java/me/earthme/luminol/config/ConfigManager.java b/src/main/java/me/earthme/luminol/config/ConfigManager.java
index 7a9907727fb221e06ae1bb0a553ef96e94b6702c..d29431cbe0c6660242615b428cfffedd3b24acd3 100644
--- a/src/main/java/me/earthme/luminol/config/ConfigManager.java
+++ b/src/main/java/me/earthme/luminol/config/ConfigManager.java
@@ -35,6 +35,7 @@ public class ConfigManager {
                 "fun.bm.lophine.carpet.config.modules"
         ));
         // Lophine end - add carpet config
+        registerConfig(builder.of("tumble", "org.tumblemc.tumble.config.modules")); //Tumble - add tumble config
         preLoad();
     }
 
